In [2]:
import pandas as pd

df = pd.read_csv("../data/data_for_model_building.csv")

df.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,...,Churn,Contract_One year,Contract_Two year,InternetService_Fiber optic,InternetService_No,PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check,MultipleLines_No phone service,MultipleLines_Yes
0,1,0,1,0,1,0,0,1,0,0,...,No,False,False,False,False,False,True,False,True,False
1,0,0,0,0,34,1,1,0,1,0,...,No,True,False,False,False,False,False,True,False,False
2,0,0,0,0,2,1,1,1,0,0,...,Yes,False,False,False,False,False,False,True,False,False
3,0,0,0,0,45,0,1,0,1,1,...,No,True,False,False,False,False,False,False,True,False
4,1,0,0,0,2,1,0,0,0,0,...,Yes,False,False,True,False,False,True,False,False,False


## Train-Test Split

In [3]:
# 1 Separate features and target:

X = df.drop("Churn", axis=1)  # Features
y = df["Churn"]               # Target

In [4]:
# 2 Split the dataset using stratified sampling to maintain the churn ratio:

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
# test_size=0.2 → 20% of data for testing.
# stratify=y → ensures churn proportion in train/test matches the original (~73/27).

In [5]:
#Check shapes:

print(X_train.shape, X_test.shape)
print(y_train.value_counts(normalize=True))
print(y_test.value_counts(normalize=True))

(5625, 24) (1407, 24)
Churn
No     0.734222
Yes    0.265778
Name: proportion, dtype: float64
Churn
No     0.734186
Yes    0.265814
Name: proportion, dtype: float64


## Data Splitting and Stratification

Now that the features are engineered, we must divide our data into a **Training Set** (to teach the model) and a **Test Set** (to evaluate its performance on unseen data).

---

## 1. Feature and Target Separation

First, we isolate the independent variables ($X$) from our dependent target variable ($y$).

```python
# Separate features and target
X = df.drop("Churn", axis=1)  # Features (everything except Churn)
y = df["Churn"]               # Target (the outcome we want to predict)

```

---

## 2. The Importance of Stratified Splitting

Since our dataset is imbalanced (~73% Stayed vs. ~27% Churned), a random split might accidentally put too many churners in the test set and not enough in the training set. To prevent this, we use **Stratified Sampling**.

```python
from sklearn.model_selection import train_test_split

# Splitting the dataset
X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.2, 
    random_state=42, 
    stratify=y
)

```

* **test_size=0.2**: We reserve 20% of the data (1,407 rows) for the final exam (testing) and use 80% (5,625 rows) for study (training).
* **stratify=y**: This ensures that both the training and test sets maintain the exact same 73/27 ratio of churners.

---

## 3. Data Split Verification

After running the split, we verify the shapes and the distribution to ensure the logic held up:

| Dataset | Row Count | Column Count | Churn Rate (%) |
| --- | --- | --- | --- |
| **Training Set** | 5,625 | 24 | 26.58% |
| **Test Set** | 1,407 | 24 | 26.58% |

**Result:** The stratified split worked exactly as intended. Both sets are mathematically representative of the original population, which is critical for building a model that generalizes well to real-world scenarios.

---

##  Analytical Perspective

By maintaining the churn ratio, we ensure the model sees enough "Churn=1" examples during training to recognize the patterns we identified in the EDA phase. Without stratification, the model might struggle to learn the high-risk profiles (Month-to-month, Fiber Optic, etc.) effectively.

---


## Model Selection & Baseline

Before jumping into complex algorithms, we need a baseline model.

**Why a baseline?**


*   Establishes a minimum benchmark to beat.

*   Helps us understand if more sophisticated models are truly adding value.

Recommended baseline for churn:
1. Logistic Regression → interpretable, fast, and suitable for binary classification.
2. We’ll evaluate with metrics sensitive to imbalance: Precision, Recall, F1-score, not just accuracy. 




In [6]:
import pickle

# Save the train/test splits for use in model evaluation
with open("../data/train_test_splits.pkl", "wb") as f:
    pickle.dump((X_train, X_test, y_train, y_test), f)

print("Train/test splits saved successfully!")

Train/test splits saved successfully!
